[WIP] Add Cake Build into the mix

gep13 opened the pull request:
Opening this for discussion.  Still work to be done here, but the basics are in place.

gep13 committed:
Added bootstrappers and empty build.cake file

gep13 committed:
Added ability to debug in VSCode

gep13 committed:
Added Cake specific ignores

gep13 committed:
Updated bootstrappers

- To not use Cake Version in tools folder

gep13 committed:
Added appveyor and travis config files

gep13 committed:
Added basic build script

- Clean, Restore, Build
- Top level tasks for AppVeyor and Travis

gep13 commented:
@onovotny let me know what you think.

Can you also list exactly what you are looking to be included?  You mentioned GitVersion, and OpenCover.  Anything else?

onovotny commented:
Looks good so far!

I was thinking things like you would do end-to-end, so perhaps GitVersion, and GitReleaseManager as well as OpenCover? Like if someone wanted to start a new project, this would be the template repo to clone and "insert solution here" type of thing.

gep13 committed:
Add installation of nuget to bootstrappers

gep13 committed:
Use project.json in favour of nuspec

gep13 committed:
Include common SolutionInfo.cs file

- Removed duplicate attributes from AssemblyInfo.cs

gep13 committed:
Added a Package step

- Including using GitVersion to assert semantic Version Number, based on project git history

gep13 commented:
@onovotny ok, I have added GitVersion and NuGet Package creation into the mix.

Can you take a look, and let me know what you think?

Also, can you add a Tests project, and I can take a look at adding OpenCover into the mix as well.  Would you want to include a step for publishing OpenCover results to Coveralls.io?

onovotny commented:
Just added a unit test project.

Code coverage with coveralls would be slick.

One suggestion would be to use nuspec as I think it gives more flexibility for creating packages while specifying a version, etc. 

Thoughts?

gep13 commented:
>@onovotny said...
Just added a unit test project.

Ok cool, I will take a look.

>@onovotny said...
One suggestion would be to use nuspec as I think it gives more flexibility for creating packages while specifying a version, etc.

As of right now, the .net core tooling doesn't support passing a nuspec file for doing the pack.  We could fall back to doing a nuget pack.

shiftkey commented on build.cake:
```diff
@@ -0,0 +1,115 @@
+var configuration   = Argument<string>("configuration", "Release");
+
+// Load other scripts.
+#load "./build/parameters.cake"
```
This file doesn't exist in the template - is it essential to the `PARAMETERS` section below?

gep13 replied:
@shiftkey yes, it is a requirement. Must be excluded through gitignore. I have it locally on my machine. I will correct later once at a computer. Thanks!

shiftkey replied:
> Must be excluded through gitignore.

I'm curious about this bit - it looks like `BuildParameters` is defined in this script, so having it ignored would affect the first-start experience after cloning a repository. 

> `error CS0246: The type or namespace name 'BuildParameters' could not be found (are you missing a using directive or an assembly reference?)`

Anyway, I'll see what is inside this file before speculating further :metal:

gep13 replied:
@shiftkey it certainly wasn't meant to be ignored. Perhaps the choice of folder name `build` was included in the default gitignore file that @onovotny created. I didn't check. I borrowed the folder structure from another repo that I worked on. The intention would certainly be for it to work straight away after doing the clone.

gep13 replied:
@shiftkey sorry for the slight delay, got caught up with a few things 😢 

I have now pushed the two files that were missing from this PR, and you should now be able to build the project.

gep13 committed:
Added missing files

gep13 commented:
@onovotny @shiftkey the more I think about this, the more I think I am duplicating effort.  The things that are being suggested to be added to this repo template are already handled by the Cake.Recipe project: https://github.com/cake-contrib/Cake.Recipe and you can see an example project that uses that here: https://github.com/gep13/Cake.Recipe.Tests  Also, there is an example output of the build from AppVeyor here: https://ci.appveyor.com/project/GaryEwanPark/cake-recipe-tests/build/0.1.0.build.5

In addition, we have an open issue to add support for .Net Core projects: https://github.com/cake-contrib/Cake.Recipe/issues/10

I am inclined to move the .Net Core projects that you have in this repo into the Cake.Recipe.Tests repository, and then concentrate the efforts of creating a generic Cake Build Script in one place.  What do you think?

To give you an idea of what Cake.Recipe does, it currently includes the following:

- Show-Info
  - Prints out some information about the current build, i.e. Target, Configuration, File Paths, etc
- Print-AppVeyor-Environment-Variables
  - Using the AppVeyorProvider within Cake, print out all the relevant information about the AppVeyor Build, very useful for debugging buids
- Clean
  - Cleans the bin/obj folders of projects, and also the build output directory
- Restore
  - NuGet Restore, based on Solution
- Build
  - Run MSBuild on the solution that is provided
  - As part of this, the individual project output are copied to a convention based folder structure, which allows for additional processing later
- Create-NuGet-Packages
  - Based on the presence of nuspec files in a convention based folder, create nuget packages, using the build output folders from Build step
- Create-Chocolatey-Packages
  - Based on the presence of nuspec files in a convention based folder, create nuget packages, using the build output folders from Build step
- Test-NUnit
  - Based on whether there are any NUnit Dll's generated as part of the build, execute OpenCover using the NUnit Console Runner and output results to a known location
- Test-xUnit
  - Based on whether there are any XUnit Dll's generated as part of the build, execute OpenCover using the XUnit Console Runner and output results to a known location
- Test-MSTest
  - Based on whether there are any MSTest Dll's generated as part of the build, execute OpenCover using the MSTest Console Runner and output results to a known location
- Test
  - This is a top level task that takes dependencies on other tasks
  - If either of the child test tasks fail, the build fails, and an HTML report is generated
- DupFinder
  - Using JetBrains DupFinder.exe, run this analysis on project files
  - If any violations are detected, build fails, and HTML report is generated
- InspectCode
  - Using JetBrains InspectCode.exe, run this analysis on project files
  - If any violations are detected, build fails, and HTML report is generated
- Analyze
  - This is a top level task that takes dependencies on other tasks
- Package
  - This is a top level task that takes dependencies on other tasks
- Upload-AppVeyor-Artifacts
  - Based on conventions, the required artifacts, i.e. the Nuget, Chocolatey, packages, etc, are uploaded to AppVeyor
- Update-Coverage-Report
  - Assuming that a Unit Test Report is generated from OpenCover, an attempt is made to publish this to Coveralls.io
- Publish-MyGet-Packages
  - Using the same artifacts that were uploaded to AppVeyor, push all Packages to MyGet
- Publish-Chocolatey-Packages
  - Using the same artifacts that were uploaded to AppVeyor, push Chocolatey Packages to Chocolatey
  - **NOTE:** This normally only happens on a tagged master branch build
- Publish-NuGet-Packages
  - Using the same artifacts that were uploaded to AppVeyor, push NuGet Packages to Chocolatey
  - **NOTE:** This normally only happens on a tagged master branch build
- Publish-GitHub-Release
  - Using the same artifacts that were uploaded to AppVeyor, add artifacts to Release which would have been created on GitHub
  - Close milestone associated with Release
  - **NOTE:** This normally only happens on a tagged master branch build
- AppVeyor
  - This is a top level task that takes dependencies on other tasks

onovotny commented:
@gep13 I did see that repo but I wasn't sure what it was for.... Seemed like an integration test repo for Cake.Recipe :)

I do think that most of this is probably already somewhere, but that what's needed is likely a rename or additional docs. My main goal is to have a "start here"/"blank" repo that people would use to start a new project and that has the existing stuff in it documented. So, I guess that'd include the "how" of reusing the cake recipe's, not recreating everything :)

gep13 commented:
@onovotny I definitely agree that additional documentation is required for Cake.Recipe.  In it's current state, I wouldn't say it is ready for "prime" time, but it is getting close.  At which point, everything required to get up and running will be included.

Sounds like we have a common goal 😄 👍

onovotny commented:
For example, looking at that recipe tests repo, I see this:
https://github.com/gep13/Cake.Recipe.Tests/blob/develop/setup.cake

But I don't see any place where I specify which cake recipe's to use, in what order and with what parameters. It's that glue that's missing.

gep13 commented:
>@onovotny said...
But I don't see any place where I specify which cake recipe's to use, in what order and with what parameters. It's that glue that's missing.

Yip, exactly.  That definitely needs to be documented.

onovotny commented:
Couple of thoughts -- with the `Print-AppVeyor-Environment-Variables` task, that seems very dangerous...as it'd leak api keys or secure environment variables?

For .NET Core, I don't think there's much to do anymore once the new tooling comes out. It'd really just be the same as building the SLN today. Things like pack/publish could be done with the `dotnet` cli, but it could also be done with a `/t:pack` parameter to msbuild.

Likewise, specifying the assembly and package versions become much easier -- they're MSBuild variables. So you can easily set them when calling MSBuild.

onovotny commented:
I would suggest/ask that if you can provide the most minimal folder structure and settings/parameters/script to use/tweak/configure Cake.Recipe, we can use that as a starting point and then submit PR's back to the recipe's as we discover places to improve (perhaps the opencover exclusions, etc).

gep13 commented:
>@onovotny said....
Couple of thoughts -- with the Print-AppVeyor-Environment-Variables task, that seems very dangerous...as it'd leak api keys or secure environment variables?

Actually, it is only the variables here that are being printed out:

https://www.appveyor.com/docs/environment-variables/

None of which are sensitive I don't think, but which can be useful during debugging.

>@onovotny said....
For .NET Core, I don't think there's much to do anymore once the new tooling comes out. It'd really just be the same as building the SLN today. Things like pack/publish could be done with the dotnet cli, but it could also be done with a /t:pack parameter to msbuild.

Yes, I agree.  I wouldn't envision all the stated tasks above being required for a .Net Core project.  Rather, it would be possible to construct which tasks you want to execute, but also include a pre-defined set of tasks for both .Net and .Net Core projects.

>@onovotny said...
Likewise, specifying the assembly and package versions become much easier -- they're MSBuild variables. So you can easily set them when calling MSBuild.

Yes, same as above.

gep13 commented:
>@onovotny said...
I would suggest/ask that if you can provide the most minimal folder structure and settings/parameters/script to use/tweak/configure Cake.Recipe, we can use that as a starting point and then submit PR's back to the recipe's as we discover places to improve (perhaps the opencover exclusions, etc).

Sounds like a plan :+1:

Tonight is a Chocolatey night for me, got a few things I need to take care of over there, but I should be able to swing back around this tomorrow.
